hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/analysis/HoodieAnalysis.scala

```diff
@@ -181,6 +183,67 @@ object HoodieAnalysis extends SparkAdapterSupport {
    */
   case class AdaptIngestionTargetLogicalRelations(spark: SparkSession) extends Rule[LogicalPlan] {
 
+    /**
+     * The function updates the HadoopFSRelation so that it uses HoodieFileIndex instead of
+     * HoodieFileIndexTimestampKeyGen. Also the data type for output attributes of the plan are changed
+     * accordingly. HoodieFileIndexTimestampKeyGen is used by HoodieBaseRelation for reading tables with
+     * Timestamp or custom key generator.
+     */
+    private def transformReaderFSRelation(logicalPlan: Option[LogicalPlan]): Option[LogicalPlan] = {
```

Comment:
How often is this function called during the lifecycle of a query?

Reply:
I can see it getting called four times. Twice for target table and twice for query fields in MergeIntoTable for example. The modification happens only once though for targetTable.

Reply:
That's an overhead right. Can we reduce the number of calls or perhaps only guard it to happen that many times only under particular conditions?

Reply:
Made some changes to make it clearer when the transformation happens. Currently transformation happens only if `useStringTypeForCustomTimestampPartition` is true for file index.